### PR TITLE
Extend conversion API with option to inject virtio drivers

### PR DESCRIPTION
**Pull request by MusicDin**
This PR adds new conversion option `virtio` (disabled by default). When enabled, LXD invokes `virt-v2v-in-place` tool on the host to inject virtio drivers into the image. For Windows images, additional drivers need to be present in `/usr/share/virt-tools`.

This requires:
- For rebase: https://github.com/canonical/lxd/pull/13385
- Custom command in LXD snap to invoke virt-v2v-in-place on the host: https://github.com/canonical/lxd-pkg-snap/pull/510

### CODE REVIEW

File: doc/howto/import_machines_to_instances.md [hunk 1]
```diff
@@ -27,13 +27,21 @@ The tool can create both containers and virtual machines:
...
-* When creating a virtual machine, you must provide a bootable disk, partition or image.
+* When creating a virtual machine, you must provide a bootable disk, partition, or an image in raw, QCOW, QCOW2, VDI, VHDX, or VMDK format.
   This means that just providing a file system is not sufficient, and you cannot create a virtual machine from a container that you are running.
   It is also not possible to create a virtual machine from the physical machine that you are using to do the migration, because the migration tool would be using the disk that it is copying.
   Instead, you could provide a bootable image, or a bootable partition or disk that is currently not in use.
 
+The tool can also inject the required drivers into the image:
+
+* Use `lxd-migrate --conversion=format,virtio` to ensure the image is converted into the raw format and virtio drivers are injected during the conversion.
+  The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where LXD server runs.
```
Comment (ru-fu) [inline doc/howto/import_machines_to_instances.md:38]:
````suggestion
* To convert the image into raw format and inject the VIRTIO drivers during the conversion, use the following command:

       lxd-migrate --conversion=format,virtio

  ```{note}
  The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where the LXD server runs.
  ```
````
To make it a bit easier to scan and find the important information.

File: doc/howto/import_machines_to_instances.md [hunk 2]
```diff
@@ -27,13 +27,21 @@ The tool can create both containers and virtual machines:
...
+* For converting Windows images, additional drivers needs to be installed on the host:
```
Comment (ru-fu) [inline doc/howto/import_machines_to_instances.md:39]:
```suggestion
* For converting Windows images, you must install additional drivers on the host:
```

File: doc/howto/import_machines_to_instances.md [hunk 3]
```diff
@@ -27,13 +27,21 @@ The tool can create both containers and virtual machines:
...
+   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) and place it in `/usr/share/virtio-win` directory.
+   * Download `pnp_wait.exe` and `rhsrvany.exe` and place them in `/usr/share/virt-tools/` directory.
```
Comment (ru-fu) [inline doc/howto/import_machines_to_instances.md:41]:
```suggestion
   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) and place it in the `/usr/share/virtio-win` directory.
   * Download `pnp_wait.exe` and `rhsrvany.exe` and place them in the `/usr/share/virt-tools/` directory.
```

File: doc/howto/import_machines_to_instances.md [hunk 4]
```diff
@@ -27,13 +27,27 @@ The tool can create both containers and virtual machines:
...
+* When creating a virtual machine, you must provide a bootable disk, partition, or an image in raw, QCOW, QCOW2, VDI, VHDX, or VMDK format.
   This means that just providing a file system is not sufficient, and you cannot create a virtual machine from a container that you are running.
   It is also not possible to create a virtual machine from the physical machine that you are using to do the migration, because the migration tool would be using the disk that it is copying.
   Instead, you could provide a bootable image, or a bootable partition or disk that is currently not in use.
 
+The tool can also inject the required drivers into the image:
+
+* To convert the image into raw format and inject the VIRTIO drivers during the conversion, use the following command:
+
+    lxd-migrate --conversion=format,virtio
```
Comment (ru-fu) [inline doc/howto/import_machines_to_instances.md:39]: This needs more indentation (should be 6 spaces I think).

File: doc/howto/import_machines_to_instances.md [hunk 5]
```diff
@@ -27,13 +27,27 @@ The tool can create both containers and virtual machines:
...
+
+  ```{note}
+  The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where the LXD server runs.
+  ```
+
+* For converting Windows images, you must additional drivers on the host:
+   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) and place it in the `/usr/share/virtio-win` directory.
```
Comment (ru-fu) [inline doc/howto/import_machines_to_instances.md:46]:
```suggestion
   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) file and place it in the `/usr/share/virtio-win` directory.
```

File: doc/.sphinx/.markdownlint/exceptions.txt [hunk 6]
```diff
@@ -1,5 +1,5 @@
-.tmp/doc/howto/import_machines_to_instances.md:108: MD034 Bare URL used
-.tmp/doc/howto/import_machines_to_instances.md:212: MD034 Bare URL used
+.tmp/doc/howto/import_machines_to_instances.md:111: MD034 Bare URL used
```
Comment (ru-fu) [inline doc/.sphinx/.markdownlint/exceptions.txt:1]: Sorry, that changed again now. ;)

File: doc/howto/import_machines_to_instances.md (hunk 2 above)
Comment (MusicDin) [reply doc/howto/import_machines_to_instances.md]: Thanks. Done :)

File: doc/howto/import_machines_to_instances.md (hunk 1 above)
Comment (MusicDin) [reply doc/howto/import_machines_to_instances.md]: Thanks.

File: doc/howto/import_machines_to_instances.md (hunk 4 above)
Comment (MusicDin) [reply doc/howto/import_machines_to_instances.md]: Fixed. Thanks :)

File: doc/howto/import_machines_to_instances.md [hunk 7]
```diff
@@ -32,8 +32,22 @@ The tool can create both containers and virtual machines:
...
+  The conversion option `virtio` requires `virt-v2v-in-place` to be installed on the host where the LXD server runs.
+  ```
+
+* For converting Windows images, you must install additional drivers on the host:
+   * Install the `virtio-win` package or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) file and place it in the `/usr/share/virtio-win` directory.
+   * Download `pnp_wait.exe` and `rhsrvany.exe` and place them in the `/usr/share/virt-tools/` directory.
+
    ````{tip}
-   If you want to convert a Windows VM from a foreign hypervisor (not from QEMU/KVM with Q35/`virtio-scsi`),
+   If you want to convert a Windows VM from a foreign hypervisor manually (not from QEMU/KVM with Q35/`virtio-scsi`),
```
Comment (MusicDin) [inline doc/howto/import_machines_to_instances.md:50]: @ru-fu 

The driver installation section (`Install the virtio-win package or download...`) repeats a few lines lower in the tip (`If you want to convert a Windows VM from a foreign hypervisor maunaly...`).

Could I just say something like this in the tip, to reduce the unnecessary repetition?

Tip title: `Expand to see how to integrate the required drivers to your Windows VM manually`
Tip content:
```
If you want to convert a Windows VM from a foreign hypervisor manually (not from 
QEMU/KVM with Q35/`virtio-scsi`), in addition to the required Windows drivers, you
must also install `virt-v2v`.

Using `virt-v2v`, you can convert images from a foreign hypervisor to `raw` images for
LXD and include the required drivers:

Example 1: ... "Rest of the content remains the same"
```

Comment (ru-fu) [reply doc/howto/import_machines_to_instances.md:50]: Good catch!

Yes, it makes sense to remove that part from the tip. Just make it very clear that you still need to follow the instructions above - something like ``If you want to convert a Windows VM from a foreign hypervisor manually (not from QEMU/KVM with Q35/`virtio-scsi`), you must install both the required Windows drivers (as described above) and `virt-v2v`.``

**Comment by tomponline** [comment]
Time to rebase please

**Comment by tomponline** [comment]
@MusicDin ready for review?

File: lxd-migrate/main_migrate.go [hunk 8]
```diff
@@ -54,7 +54,7 @@ func (c *cmdMigrate) command() *cobra.Command {
 `
 	cmd.RunE = c.run
 	cmd.Flags().StringVar(&c.flagRsyncArgs, "rsync-args", "", "Extra arguments to pass to rsync"+"``")
-	cmd.Flags().StringSliceVar(&c.flagConversionOpts, "conversion", []string{"format"}, "List of conversion opts. Allowed values are: [format]")
+	cmd.Flags().StringSliceVar(&c.flagConversionOpts, "conversion", []string{"format"}, "List of conversion opts. Allowed values are: [format, virtio]")
```
Comment (tomponline) [inline lxd-migrate/main_migrate.go:57]: Can we change "List of conversion opts" to "List of conversion options to apply" please

File: lxd/storage/backend_lxd.go [hunk 9]
```diff
@@ -1826,8 +1827,8 @@ func (b *lxdBackend) imageConversionFiller(imgPath string, imgFormat string) fun
 
 		// Convert uploaded image from backups directory into RAW format on the instance volume.
 		cmd := []string{
-			// Use prlimit to limit QEMU to 1 GiB address space and 120 seconds of CPU time.
-			"prlimit", "--cpu=120", "--as=1073741824",
+			// Use prlimit to limit QEMU to 1 GiB address space and 5 minutes of CPU time.
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:1830]: this has been bumped but the commit message nor comment explains why.

File: lxd/storage/backend_lxd.go [hunk 10]
```diff
@@ -2549,6 +2531,51 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
...
+			return err
+		}
+
+		imgFormat, _, err := b.qemuImageInfo(diskPath)
+		if err != nil {
+			return err
+		}
+
+		// Ensure the disk path points to a disk in raw format.
+		if imgFormat != "raw" {
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2558]: why is this check needed?

Should we have not already converted the volume to raw by now?

File: lxd/storage/backend_lxd.go [hunk 11]
```diff
@@ -2558,6 +2585,35 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
 	return nil
 }
 
+// qemuImageInfo retrieves the format and size of an image on the given path.
+func (b *lxdBackend) qemuImageInfo(imagePath string) (format string, bytes int64, err error) {
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2589]: Should ImageUnpack be updated to use this?

File: doc/howto/import_machines_to_instances.md [hunk 12]
```diff
@@ -32,18 +32,28 @@ The tool can create both containers and virtual machines:
   It is also not possible to create a virtual machine from the physical machine that you are using to do the migration, because the migration tool would be using the disk that it is copying.
   Instead, you could provide a bootable image, or a bootable partition or disk that is currently not in use.
 
+The tool can also inject the required drivers into the image:
```
Comment (tomponline) [inline doc/howto/import_machines_to_instances.md:35]: required *virtio* drivers

File: lxd/storage/backend_lxd.go (hunk 9 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:1830]: Will add the comment. The reason is that large VMs such as Windows can take longer then 2 minutes to convert.

File: lxd/storage/backend_lxd.go (hunk 10 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:2558]: Correct. This should never occur. Will remove

File: lxd/storage/backend_lxd.go [hunk 13]
```diff
@@ -2549,6 +2531,51 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
...
+			return err
+		}
+
+		// Ensure the disk path points to a disk in raw format.
+		if imgFormat != "raw" {
+			return fmt.Errorf(`Expected volume format for driver injection is "raw", but got %q`, imgFormat)
+		}
+
+		// Run virt-v2v-in-place to inject virtio drivers.
+		cmd := exec.Command(
```
[the remark this answers is not in the document]
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:2563]: Good point. I'll tests that out and fix if needed

File: lxd/storage/backend_lxd.go [hunk 14]
```diff
@@ -2549,6 +2531,51 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
...
+			// Use prlimit to limit QEMU to 1 GiB address space and 5 minutes of CPU time.
+			"prlimit", "--cpu=300", "--as=1073741824",
+			"virt-v2v-in-place", "-i", "disk", "-if", "raw", "--block-driver", "virtio-scsi", diskPath,
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2566]: We should run this as a low priority, like we do with qemu-img convert using 

"nice", "-n19", // Run with low priority to reduce CPU impact on other processes.

Comment (MusicDin) [reply lxd/storage/backend_lxd.go:2566]: For `qemu-img convert` we also use `prlimit`. Should we also run the `qemu-img convert` with low priority instead of limiting it to 5 minutes?

Comment (tomponline) [reply lxd/storage/backend_lxd.go:2566]: both please

File: doc/howto/import_machines_to_instances.md [hunk 15]
```diff
@@ -32,18 +32,28 @@ The tool can create both containers and virtual machines:
...
-   <summary>Expand to see how to integrate the required drivers to your Windows VM</summary>
-   Install the required tools on the host:
+   If you want to convert a Windows VM from a foreign hypervisor manually,
+   you must install both the required Windows drivers (as described above) and `virt-v2v` (>= 2.3.4).
 
-   1. Install `virt-v2v` version >= 2.3.4 (this is the minimal version that supports the `--block-driver` option).
-   1. Install the `virtio-win` package, or download the [`virtio-win.iso`](https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/stable-virtio/virtio-win.iso) image and put it into the `/usr/share/virtio-win` folder.
-   1. You might also need to install [`rhsrvany`](https://github.com/rwmjones/rhsrvany).
+   <details>
+   <summary>Expand to see how to convert your Windows VM using `virt-v2v`</summary>
```
[the remark this answers is not in the document]
Comment (MusicDin) [reply doc/howto/import_machines_to_instances.md:54]: Thanks, probably this is also the reason why it detects `virt` as missing in dictionary.

[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/howto/import_machines_to_instances.md:54]: Yes - code style is automatically excluded from the spelling check, but since the markup didn't work here, it raised an error.

File: lxd/storage/backend_lxd.go (hunk 14 above)
Comment (tomponline) [reply lxd/storage/backend_lxd.go:2556]: As discussed lets use prlimit just for the qemu-img info call. Ta

**Comment by MusicDin** [comment]
@tomponline Sorry, forgot to mark as ready for review

File: lxd/storage/backend_lxd.go [hunk 16]
```diff
@@ -1840,6 +1840,12 @@ func (b *lxdBackend) imageConversionFiller(imgPath string, imgFormat string) fun
 			return -1, fmt.Errorf("qemu-img convert: failed to convert image from %q to %q format: %v", imgFormat, "raw", err)
 		}
 
+		// Remove the image after the conversion to free up the space as soon as possible.
+		err = os.Remove(imgPath)
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:1846]: Is the file also removed on error?

File: lxd/storage/backend_lxd.go [hunk 17]
```diff
@@ -1828,7 +1830,7 @@ func (b *lxdBackend) imageConversionFiller(imgPath string, imgFormat string) fun
 		cmd := []string{
 			// Run with low priority to reduce CPU impact on other processes.
 			"nice", "-n19",
-			"qemu-img", "convert", "-f", imgFormat, "-O", "raw", imgPath, diskPath,
+			"qemu-img", "convert", "-f", imgFormat, "-O", "raw", imgPath, diskPath, "-t", "writeback",
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:1833]: this should be a separate commit as not related to commit message

File: lxd/storage/backend_lxd.go [hunk 18]
```diff
@@ -2476,40 +2478,23 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
...
-			// Use prlimit because qemu-img can consume considerable RAM & CPU time if fed
-			// a maliciously crafted disk image. Since cloud tenants are not to be trusted,
-			// ensure QEMU is limited to 1 GiB address space and 2 seconds of CPU time.
-			// This should be more than enough for real world images.
-			"prlimit", "--cpu=2", "--as=1073741824",
-			"qemu-img", "info", imgPath, "--output", "json",
-		}
-
-		out, err := apparmor.QemuImg(b.state.OS, cmd, imgPath, "")
+		imgFormat, imgBytes, err := qemuImageInfo(b.state.OS, imgPath)
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2481]: can this be a separate commit?

File: lxd/storage/backend_lxd.go [hunk 19]
```diff
@@ -2555,6 +2540,81 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
...
+		defer func() { _, _ = b.driver.UnmountVolume(vol, true, op) }()
+
+		diskPath, err := b.driver.GetVolumeDiskPath(vol)
+		if err != nil {
+			return err
+		}
+
+		out, err := exec.Command("virt-v2v-in-place", "--version").CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("Failed to get virt-v2v-in-place version (%v): %s", err, string(out))
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2563]: Please use %w to wrap the error

File: lxd/storage/backend_lxd.go [hunk 20]
```diff
@@ -2555,6 +2540,81 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
...
+		}
+
+		// checkMinVersions extracts version numbers and compares them one by one returning
+		// true if given version is larger then or equal to minimum version and false otherwise.
+		checkMinVersion := func(version string, min string) bool {
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2568]: Can you use `version.NewDottedVersion()` for this comparison?

File: lxd/storage/backend_lxd.go [hunk 21]
```diff
@@ -2555,6 +2540,81 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
...
+		cmd := exec.Command(
+			// Run with low priority to reduce the CPU impact on other processes.
+			"nice", "-n19",
+			"virt-v2v-in-place", "-i", "disk", "-if", "raw", "--block-driver", "virtio-scsi", diskPath,
+		)
+
+		// Instruct virt-v2v-in-place where to search for windows drivers.
+		cmd.Env = append(os.Environ(),
+			"VIRTIO_WIN=/usr/share/virtio-win/virtio-win.iso",
+			"VIRT_TOOLS_DATA_DIR=/usr/share/virt-tools")
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2609]: Please can the closing bracket be on a new line

File: lxd/storage/backend_lxd.go [hunk 22]
```diff
@@ -2555,6 +2540,81 @@ func (b *lxdBackend) CreateInstanceFromConversion(inst instance.Instance, conn i
...
+
+		out, err = cmd.CombinedOutput()
+		if err != nil {
+			b.logger.Error("Image conversion failed", logger.Ctx{"error": string(out)})
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:2613]: Do we need the logger if we are returning the error? Generally we dont do this sort of duplication.

**Review by tomponline — COMMENTED** [review]
A few nits, but i would say very nearly ready.

File: lxd/storage/backend_lxd.go (hunk 16 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:1846]: No, as far as I know if the error occurs the file removal failed.

If the question is whether we should ignore the error:

We can ignore the error, since at this point conversion was successful, so it is a bit of a waste to revert everything just because the temporary file was not successfully deleted.

Or, we can error out to let the user know that there is some file left over. I've never reached this error, but if error occurs we will at least be aware of it (aiming at permission - the file has 0600 permission).

File: lxd/storage/backend_lxd.go (hunk 20 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:2568]: I knew there must already be something for comparing versions. Thanks :)

File: lxd/storage/backend_lxd.go (hunk 16 above)
Comment (tomponline) [reply lxd/storage/backend_lxd.go:1846]: I meant if an error occurs earlier than this, does the file still get removed?

File: lxd/storage/backend_lxd.go (hunk 22 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:2613]: If virt-v2v-in-place fails the error will be `exit status 1`, so I wanted to just log the entire output of the virt-v2v-in-place. We can also return it as error, but it did not fell right to return log within error. Maybe there is a better way, any thoughts?

File: lxd/storage/backend_lxd.go (hunk 16 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:1846]: Ahh, yes. Before the imageConversionFiller is invoked, the file is created in `createInstanceFromConversion` and there is a `defer` that cleans the file up.

File: lxd/storage/backend_lxd.go (hunk 22 above)
Comment (tomponline) [reply lxd/storage/backend_lxd.go:2613]: Can you wrap the error with "Image conversion failed: %w"?

**Comment by MusicDin** [comment]
@tomponline Ready for another round of review

File: lxd/storage/backend_lxd.go [hunk 23]
```diff
@@ -1828,7 +1828,7 @@ func (b *lxdBackend) imageConversionFiller(imgPath string, imgFormat string) fun
 		cmd := []string{
 			// Run with low priority to reduce CPU impact on other processes.
 			"nice", "-n19",
-			"qemu-img", "convert", "-f", imgFormat, "-O", "raw", imgPath, diskPath,
+			"qemu-img", "convert", "-f", imgFormat, "-O", "raw", imgPath, diskPath, "-t", "writeback",
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:1833]: Can you attribute this to https://github.com/lxc/incus/pull/1015/commits/bbc1ea8cd0cb01b4830fd960007f4c5b6d05edf2 please by way of a cherry-pick

**Review by tomponline — CHANGES_REQUESTED** [review]
Just an issue with correct attribution of the writeback change.

File: lxd/storage/backend_lxd.go (hunk 23 above)
Comment (MusicDin) [reply lxd/storage/backend_lxd.go:1833]: Done

**Review by tomponline — APPROVED** [review]
LGTM!

**Comment by tomponline** [comment]
@MusicDin cool thanks, will merge this soon. Do you have a PR for lxd-ci to test virtio mode in the works?
